### PR TITLE
Improvement: Improvement svg curve draw

### PR DESCRIPTION
**Pull request by MaShizhen**

### CODE REVIEW

File: src/app/flux/editor/index.js [hunk 1]
```diff
@@ -581,6 +586,7 @@ export const actions = {
      * @param modelID - optional, used in project recovery
      */
     generateModel: (headType, { originalName, uploadName, sourceWidth, sourceHeight, mode, sourceType, config, gcodeConfig, transformation, modelID, zIndex, isLimit }) => (dispatch, getState) => {
+        console.log('>> generateModel');
```
Comment (xiekun1992) [inline src/app/flux/editor/index.js:589]: remove log

File: src/app/flux/operation-history/index.js [hunk 2]
```diff
@@ -38,6 +42,21 @@ export const actions = {
         }
         dispatch(projectActions.autoSaveEnvironment(headType));
     },
+    isDrawOperation(operation) {
+        return operation instanceof DrawLine || operation instanceof DrawDelete || operation instanceof DrawTransform || operation instanceof DrawStart;
+    },
+    clearDrawOperations: (headType, keeyDrawStart) => (dispatch, getState) => {
+        const { history } = getState()[headType];
+        history.history = history.history.filter((item) => {
```
Comment (xiekun1992) [inline src/app/flux/operation-history/index.js:50]: do not edit access `history.history` directly, expose API from `OperationHistory`

File: src/app/models/SvgModel.js [hunk 3]
```diff
@@ -475,6 +476,17 @@ class SvgModel extends BaseModel {
             //     numberAttrs.push('x1', 'y1', 'x2', 'y2');
             //     break;
             case 'path': {
+                const isDraw = elem.getAttribute('id')?.includes('graph');
+                if (isDraw && config['svg-path-d']) {
+                    elem.setAttribute('d', config['svg-path-d']);
+                    elem.setAttribute('source', config['svg-path-d']);
+
+                    const bbox = elem.getBBox();
+                    elem.setAttribute('x', (bbox.x + bbox.width / 2).toString());
```
Comment (xiekun1992) [inline src/app/models/SvgModel.js:485]: use ES6 Template Strings

File: src/app/ui/SVGEditor/svg-content/DrawGroup/DrawGroup.ts [hunk 4]
```diff
@@ -0,0 +1,788 @@
...
+            }
+            // move line
+            if (leftKeyPressed && this.selected.line) {
+                this.transformLine(dx, dy);
+                this.operationGroup.updateOperation(this.selected.line.ele);
+            }
+        }
+    }
+
+    private compD() {
```
Comment (xiekun1992) [inline src/app/ui/SVGEditor/svg-content/DrawGroup/DrawGroup.ts:676]: use a full name

File: src/app/ui/SVGEditor/svg-content/SVGContentGroup.js [hunk 5]
```diff
@@ -37,6 +47,30 @@ class SVGContentGroup {
             scale: this.scale
         });
         this.operatorPoints.showOperator(true);
+
+        this.drawGroup = new DrawGroup(this.group);
+
+        this.drawGroup.onDrawLine = (line) => {
+            this.onDrawLine(line);
```
Comment (xiekun1992) [inline src/app/ui/SVGEditor/svg-content/SVGContentGroup.js:54]: declare `onDrawXXX` methods explicitly, these methods area dynamically assigned which can be confused

**Review by xiekun1992 — CHANGES_REQUESTED** [review]

File: src/app/flux/editor/index.js [hunk 6]
```diff
@@ -1312,18 +1322,23 @@ export const actions = {
...
         dispatch(actions.resetProcessState(headType));
     },
 
-    selectAllElements: (headType) => (dispatch, getState) => {
-        const { SVGActions } = getState()[headType];
-        // dispatch(actions.clearSelection(headType));
-        SVGActions.selectAllElements();
-        dispatch(baseActions.render(headType));
+    selectAllElements: (headType) => async (dispatch, getState) => {
+        const { SVGActions, SVGCanvasMode, SVGCanvasExt } = getState()[headType];
```
Comment (jane-rose) [inline src/app/flux/editor/index.js:1332]: Option: disable select all element shorcut when in draw mode

File: src/app/flux/laser/index.js [hunk 7]
```diff
@@ -367,7 +373,9 @@ export default function reducer(state = INITIAL_STATE, action) {
     if (headType === 'laser') {
         switch (type) {
             case ACTION_UPDATE_STATE: {
-                return Object.assign({}, state, { ...action.state });
+                const ll = Object.assign({}, state, { ...action.state });
```
Comment (xiekun1992) [inline src/app/flux/laser/index.js:376]: restore this after debug

File: src/shared/lib/utils.js [hunk 8]
```diff
@@ -1,4 +1,5 @@
 import _ from 'lodash';
+import { JudgeMoveMiniDistance } from '../../app/constants';
```
Comment (xiekun1992) [inline src/shared/lib/utils.js:2]: `JudgeMoveMiniDistance` => `JUDGE_MOVE_MINI_DISTANCE`

File: src/app/ui/SVGEditor/svg-content/DrawGroup/CursorGroup.ts [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+        const cursor: SVGSVGElement = createSVGElement({
+            element: 'svg',
+            attr: {
+                width: cursorWidth / this.scale,
+                height: cursorWidth / this.scale,
+                viewBox: '0 0 90 90',
+                fill: 'none'
+            }
+        });
+        cursor.innerHTML = `
```
Comment (xiekun1992) [inline src/app/ui/SVGEditor/svg-content/DrawGroup/CursorGroup.ts:83]: refer: [cursor](https://developer.mozilla.org/en-US/docs/Web/CSS/cursor), use CSS to modify cursor

**Review by xiekun1992 — CHANGES_REQUESTED** [review]

**Pull request merged**